### PR TITLE
#27  | Order service

### PR DESCRIPTION
**Pull request by SavinaMih**

### CODE REVIEW

File: src/Services/Catalog/src/Catalog.Persistence/Catalog.Persistence.csproj [hunk 1]
```diff
@@ -16,11 +16,19 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="BuildingBlocks.Common" Version="1.0.0.65" />
```
Comment (MartsTech) [inline src/Services/Catalog/src/Catalog.Persistence/Catalog.Persistence.csproj:19]: This packages does not belog here at all. Why change the catalog service?

File: src/Services/Catalog/src/Catalog.Domain/Catalog.Domain.csproj [hunk 2]
```diff
@@ -6,4 +6,15 @@
     <Nullable>enable</Nullable>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="BuildingBlocks.Common" Version="1.0.0.65" />
```
Comment (MartsTech) [inline src/Services/Catalog/src/Catalog.Domain/Catalog.Domain.csproj:10]: This packages does not belog here at all. Why change the catalog service?

File: src/Services/Catalog/src/Catalog.Application/Catalog.Application.csproj [hunk 3]
```diff
@@ -10,6 +10,17 @@
     <None Remove="Products\GetProducts\GetAllProductsQuery.cs~RF2b105e86.TMP" />
   </ItemGroup>
 
+  <ItemGroup>
+    <PackageReference Include="BuildingBlocks.Common" Version="1.0.0.65" />
```
Comment (MartsTech) [inline src/Services/Catalog/src/Catalog.Application/Catalog.Application.csproj:14]: This packages does not belog here at all. Why change the catalog service?

File: src/Services/Catalog/src/Catalog.Api/Catalog.Api.csproj [hunk 4]
```diff
@@ -8,10 +8,18 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="BuildingBlocks.Common" Version="1.0.0.65" />
```
Comment (MartsTech) [inline src/Services/Catalog/src/Catalog.Api/Catalog.Api.csproj:11]: This packages does not belog here at all. Why change the catalog service?

File: src/Services/Bookmarks/src/Bookmarks.Persistence/Bookmarks.Persistence.csproj [hunk 5]
```diff
@@ -7,15 +7,23 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="BuildingBlocks.Common" Version="1.0.0.65" />
```
Comment (MartsTech) [inline src/Services/Bookmarks/src/Bookmarks.Persistence/Bookmarks.Persistence.csproj:10]: This packages does not belog here at all. Why change the bookmarks service?

File: src/Services/Bookmarks/src/Bookmarks.Domain/Bookmarks.Domain.csproj [hunk 6]
```diff
@@ -6,4 +6,15 @@
     <Nullable>enable</Nullable>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="BuildingBlocks.Common" Version="1.0.0.65" />
```
Comment (MartsTech) [inline src/Services/Bookmarks/src/Bookmarks.Domain/Bookmarks.Domain.csproj:10]: This packages does not belog here at all. Why change the bookmarks service?

File: src/Services/Bookmarks/src/Bookmarks.Application/Bookmarks.Application.csproj [hunk 7]
```diff
@@ -6,6 +6,17 @@
     <Nullable>enable</Nullable>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="BuildingBlocks.Common" Version="1.0.0.65" />
```
Comment (MartsTech) [inline src/Services/Bookmarks/src/Bookmarks.Application/Bookmarks.Application.csproj:10]: This packages does not belog here at all. Why change the bookmarks service?

File: src/Services/Bookmarks/src/Bookmarks.Api/Bookmarks.Api.csproj [hunk 8]
```diff
@@ -7,11 +7,19 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="BuildingBlocks.Common" Version="1.0.0.65" />
```
Comment (MartsTech) [inline src/Services/Bookmarks/src/Bookmarks.Api/Bookmarks.Api.csproj:10]: This packages does not belog here at all. Why change the bookmarks service?

File: src/BuildingBlocks/BuildingBlocks.csproj [hunk 9]
```diff
@@ -7,19 +7,24 @@
     </PropertyGroup>
 
     <ItemGroup>
+      <PackageReference Include="BuildingBlocks.Common" Version="1.0.0.65" />
```
Comment (MartsTech) [inline src/BuildingBlocks/BuildingBlocks.csproj:10]: There should have been no need for additional packages. At least Microsoft.Tpl.Dataflow is very unnecessary

File: src/ApiGateway/src/ApiGateway.csproj [hunk 10]
```diff
@@ -8,6 +8,14 @@
     </PropertyGroup>
 
     <ItemGroup>
+      <PackageReference Include="BuildingBlocks.Common" Version="1.0.0.65" />
```
Comment (MartsTech) [inline src/ApiGateway/src/ApiGateway.csproj:11]: This packages does not belog here at all. Why change the api-gateway service?

File: OrderService.Api/OrderLists/GetOrderList/GetOrderListController.cs [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+{
+    public class GetWishlistsController : UserController
+    {
+        [HttpGet]
+        public async Task<IActionResult> GetAllOrderists()
+        {
+            return HandleResult(await Mediator.Send(new GetAllOrderListsQuery.Query(false)));
+        }
+
+        [HttpGet("{includeBookmarks:bool}")]
```
Comment (MartsTech) [inline OrderService.Api/OrderLists/GetOrderList/GetOrderListController.cs:15]: There are no bookmarks in this service

File: OrderService.Api/OrderLists/GetOrderList/GetOrderListController.cs [hunk 12]
```diff
@@ -0,0 +1,27 @@
+﻿using OrderService.Application.OrderLists.GetOrderList;
+using BuildingBlocks.Web;
+using Microsoft.AspNetCore.Mvc;
+
+namespace OrderService.Api.OrderLists.GetOrderList
+{
+    public class GetWishlistsController : UserController
+    {
+        [HttpGet]
+        public async Task<IActionResult> GetAllOrderists()
```
Comment (MartsTech) [inline OrderService.Api/OrderLists/GetOrderList/GetOrderListController.cs:10]: Typo

File: OrderService.Api/OrderLists/GetOrderList/GetOrderListController.cs [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+        {
+            return HandleResult(await Mediator.Send(new GetAllOrderListsQuery.Query(false)));
+        }
+
+        [HttpGet("{includeBookmarks:bool}")]
+        public async Task<IActionResult> GetAllWishlists(bool includeProducts)
```
Comment (MartsTech) [inline OrderService.Api/OrderLists/GetOrderList/GetOrderListController.cs:16]: Wishlists and includeProducts?

File: OrderService.Api/OrderLists/AddOrderList/AddOrderListController.cs [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+using Microsoft.AspNetCore.Mvc;
+
+namespace OrderService.Api.OrderLists.AddOrderList
+{
+    public class AddOrderListController : UserController
+    {
+        [HttpPost]
+        public async Task<IActionResult> AddOrderList([FromForm] AddOrderListInput input)
+        {
+            Result<WishlistDto> result = await Mediator.Send(new AddOrderListCommand.Command(input));
```
Comment (MartsTech) [inline OrderService.Api/OrderLists/AddOrderList/AddOrderListController.cs:15]: WishlistDto?

File: OrderService.Api/OrderLists/GetOrderList/GetOrderListController.cs [hunk 15]
```diff
@@ -0,0 +1,27 @@
+﻿using OrderService.Application.OrderLists.GetOrderList;
+using BuildingBlocks.Web;
+using Microsoft.AspNetCore.Mvc;
+
+namespace OrderService.Api.OrderLists.GetOrderList
+{
+    public class GetWishlistsController : UserController
```
Comment (MartsTech) [inline OrderService.Api/OrderLists/GetOrderList/GetOrderListController.cs:7]: GetWishlistController?

File: OrderService.Api/OrderLists/DeleteOrderList/DeleteOrderListConsumer.cs [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+        private readonly ILogger<DeleteOrderListConsumer> _logger;
+
+        public DeleteOrderListConsumer(ILogger<DeleteOrderListConsumer> logger)
+        {
+            _logger = logger;
+        }
+
+        public Task Consume(ConsumeContext<DeleteOrderListDeleted> context)
+        {
+            _logger.LogInformation("Wishlist deleted with ID: {MessageId}", context.Message.Id);
```
Comment (MartsTech) [inline OrderService.Api/OrderLists/DeleteOrderList/DeleteOrderListConsumer.cs:17]: Wishlist?

File: OrderService.Api/Order/AddOrder/AddOrderConsumer.cs [hunk 17]
```diff
@@ -0,0 +1,20 @@
+﻿using BuildingBlocks.MassTransit.Contracts;
+using MassTransit;
+
+namespace OrderService.Api.Order.AddOrder
+
+public class AddBookmarkConsumer : IConsumer<OrderCreated>
```
Comment (MartsTech) [inline OrderService.Api/Order/AddOrder/AddOrderConsumer.cs:6]: BookmarkConsumer?

File: OrderService.Api/Program.cs [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+app.MapControllers();
+
+using var scope = app.Services.CreateScope();
+
+var services = scope.ServiceProvider;
+
+try
+{
+    var context = services.GetRequiredService<BookmarksDbContext>();
```
Comment (MartsTech) [inline OrderService.Api/Program.cs:36]: Bookmarks?

File: OrderService.Api/Program.cs [hunk 19]
```diff
@@ -0,0 +1,45 @@
+using Bookmarks.Api.Extensions;
```
Comment (MartsTech) [inline OrderService.Api/Program.cs:1]: Bookmarks???

File: OrderService.Domain/Orders/IOrderRepository.cs [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+using System.Threading.Tasks;
+
+namespace OrderService.Domain
+{
+    public interface IOrderRepository
+    {
+        Task<bool> AddOrder(Order order);
+        Task<Order?> GetOrderById(Guid id, Guid userId);
+        Task<bool> UpdateOrder(Guid id, int newQuantity, Guid userId);
+        Task<bool> Deleteorder(Guid id, Guid userId);
```
Comment (MartsTech) [inline OrderService.Domain/Orders/IOrderRepository.cs:15]: Typo

**Review by MartsTech — CHANGES_REQUESTED** [review]

File: OrderService.Application/OrderList/GetList/GetListByIdQuery.cs [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+            _userService = userService;
+        }
+
+        public async Task<Result<WishlistDto>> Handle(Query request, CancellationToken cancellationToken)
+        {
+            var userId = _userService.GetCurrentUserId();
+
+            if (userId == null)
+            {
+                return Result<WishlistDto>.Failure("No user id found");
```
Comment (bdbelchev) [inline OrderService.Application/OrderList/GetList/GetListByIdQuery.cs:37]: This is the wishlist DTO.

File: OrderService.Application/Orders/DeleteOrder/DeleteOrderCommand.cs [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+using BuildingBlocks.Core;
+using BuildingBlocks.EFCore;
+using FluentValidation;
+using FluentValidation.Results;
+using MediatR;
+using System.ComponentModel.DataAnnotations;
+
+namespace OrderService.Application.Orders.DeleteOrder
+
+public sealed class DeleteBookmarkCommand
```
Comment (bdbelchev) [inline OrderService.Application/Orders/DeleteOrder/DeleteOrderCommand.cs:18]:
```suggestion
public sealed class DeleteOrderCommand
```

File: OrderService.Domain/Orders/IOrder.cs [hunk 23]
```diff
@@ -0,0 +1,18 @@
+﻿sing System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OrderService.Domain
+{
+    public interface IBookmark
```
Comment (bdbelchev) [inline OrderService.Domain/Orders/IOrder.cs:9]:
```suggestion
    public interface IOrder
```

File: OrderService.Persistence/OrderLists/OrderListConfiguration.cs [hunk 24]
```diff
@@ -0,0 +1,30 @@
+﻿using OrderService.Domain.OrderLists;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+
+namespace Bookmarks.Persistence.Wishlists
```
Comment (bdbelchev) [inline OrderService.Persistence/OrderLists/OrderListConfiguration.cs:5]: Leftover namespace from other project.

File: OrderService.Persistence/OrderLists/OrderListRepository.cs [hunk 25]
```diff
@@ -0,0 +1,56 @@
+﻿using OrderService.Domain.OrderLists;
+using Microsoft.EntityFrameworkCore;
+
+namespace Bookmarks.Persistence.Wishlists
```
Comment (bdbelchev) [inline OrderService.Persistence/OrderLists/OrderListRepository.cs:4]: Leftover namespace.

File: OrderService.Persistence/Orders/OrderRepository.cs [hunk 26]
```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Bookmarks.Persistence.Wishlists;
```
Comment (bdbelchev) [inline OrderService.Persistence/Orders/OrderRepository.cs:6]:
```suggestion
```

File: OrderService.Persistence/OrdersDbContext.cs
Comment (bdbelchev) [inline OrderService.Persistence/OrdersDbContext.cs:1]: I'm not sure we should depend on classes from the other service. Also, this is not CatalogDBContext, and I don't see why we should add DbSets for Products.

File: OrderService.Persistence/UnitOfWork.cs [hunk 27]
```diff
@@ -0,0 +1,39 @@
+﻿using BuildingBlocks.EFCore;
+using OrderService.Persistence;
+
+namespace OrderService.Persistence;
+
+public sealed class UnitOfWork : IUnitOfWork, IDisposable
+{
+    private readonly CatalogDbContext _context;
```
Comment (bdbelchev) [inline OrderService.Persistence/UnitOfWork.cs:8]: Maybe OrdersDbContext?

**Review by bdbelchev — CHANGES_REQUESTED** [review]
In addition to the issues already pointed out, I noticed the directory structure is incorrect, as the project files and directories should be located in src/Services/...

Also, there are some leftover namespaces from the Bookmarks service, as well as some usage of its entities. A good way to check would be to remove every `using` declaration that references Bookmarks, and check if there are errors. That way you will quickly know where you are referencing classes that belong to other projects.

**Pull request closed without merging**